### ISSUE TITLE
clang tool failing for native windows clang/llvm

### ISSUE THREAD

**Issue by mwichmann** [SCons/scons#3480]
[_this has been edited slightly from the original_]
scons' clang tool doesn't find resources to compile against if the tool found the native windows clang; it does work if it found it in a mingw environment.

The "proof" part of this problem is that the clang tests in AppVeyor (Windows) CI are failing to find standard headers and other components, but only on one combination.  Currently the CI runs against Python 3.5 on a Visual Studio 2015 image, Python 2.7 and 3.6 on a Visual Studio 2017 image, and Python 3.7 on a Visual Studio 2019 image - it is this last one where the failure is seen (there are six failing tests).  The other three combos do not fail, and yes, they are reporting those tests as run, not skipped. Here is one error to illustrate the "failure to find":
```
183/294 (62.24%) C:\\Python37\\python.exe test\Clang\clang_default_environment.py
C:\projects\scons\src\script\scons.py returned 2
STDOUT =========================================================================
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
clang -o foo.obj -c foo.c
scons: building terminated because of errors.
STDERR =========================================================================
foo.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
scons: *** [foo.obj] Error 1
``` 

On to less provable parts. AppVeyor has a list of software components in each image [here](https://www.appveyor.com/docs/windows-images-software/#llvm). It looks like the build environment has the same LLVM version for all, so the issue subject line may end up being wrong.  It is possible that the tool initialization finds clang from mingw first, and mingw-w64 is no longer installed by default in the 2019 image, while it is in the 2015 and 2017 images.  Which could mean that clang does not work at all as a native Windows install, only when found in mingw.

In addition, the LLVM project now provides a binary `clang-cl` which is intended to work in the MSVC environment - it takes the `cl` style command line switches instead of the `gcc` style. SCons knows nothing about this - it won't search for `clang-cl`, and if found, it would not know it has to supply switches in a different format, nor that is has to supply the environment variables that would be discovered by running `vcvarsall.bat`.  Note: the request to support this has moved off to #3482

**Comment by mwichmann** [comment]
Exracting a full build log from AppVeyor, it looks like the guess about the other images finding an msys version of clang first might be the case. Here's a trimmed down Windows search path for the image:
```
Path=C:\Python37;
C:\Python37\Scripts;
C:\ProgramData\chocolatey\bin;
C:\MinGW\bin;
C:\MinGW\msys\1.0\bin; 
C:\cygwin\bin;
... [43 skipped]
C:\Program Files\LLVM\bin;
...
```

**Comment by bdbaddog** [comment]
Yes supporting clang-cl should be a different tool than clangcc

**Comment by mwichmann** [comment]
Okay, so then it sounds like the fix for the failing tests should be to not run the tests.  What's the best approach to that?  Add logic to the clang tool so it doesn't initialize if it only finds the Windows-native clang? If so, how would it decide it got the wrong one - given we can't guarantee the order of path elements in the general case, though we can figure out what it is for AppVeyor specifically? Or bolt something into the tests?

Should I open a feature request for a clang-cl tool?  Or is it better if that appears first as a contrib tool?

**Comment by bdbaddog** [comment]
clang should work on windows with msvc (as well as clang-cl which take msvc style command line arguments).. 

Likely the clangc is not finding msvc's header files and thus tests failing?

**Comment by mwichmann** [comment]
Indeed it's not. It needs the vcvars stuff - which are not supplied when calling clang even though they're somewhere collected if the msvc setup previously worked.

**Comment by bdbaddog** [comment]
Looks like on windows we'll have to call the msvc initialization logic from clang tool. Also add those two shell environment variables to be propagated from vcvarsall.bat's setting capture logic.

**Comment by mwichmann** [comment]
I can do the latter, wondering if there are others that are valuable as well.

**Comment by mwichmann** [comment]
I can do the latter, wondering if there are others that are valuable as well.

**Comment by bdbaddog** [comment]
I checked clang source code and those are the only two it's looking for to find MSVC install.

**Comment by bdbaddog** [comment]
This should be resolved with PR #3505
